lib/framework/procbot.ts

```diff
@@ -108,58 +118,52 @@ export class ProcBot {
 
 	/**
 	 * Add a new type of ServiceListener to the client.
-	 * Should the ServiceListener already exist on the client, this will do nothing.
+	 * Should the ServiceListener already exist on the client with the handle/name, this will do nothing.
+	 *
 	 * @param name   The name of the ServiceListener to add.
-	 * @param data?  Any relevant data required to construct the ServiceListener.
+	 * @param data?  A ServiceBase extended constructor, should data be passed to a Service constructor.
 	 * @return       The constructed ServiceListener or void should it already exist or fail.
 	 */
 	protected addServiceListener(name: string, data?: any): ServiceListener | void {
-		const service = this.getService(name);
-		let listener;
-
-		if (service && !_.find(this.listeners, [ 'serviceName', name ])) {
-			listener = service.createServiceListener(data);
-			this.listeners.push(listener);
-		}
-
-		return listener;
+		return <ServiceListener>this.addService('listener', name, data);
```

Comment:
If you really wanna go wild, here's a half-built totally type-safe promisify implementation I was working on a little while ago, which uses this _heavily_: https://github.com/pimterry/typed-promisify-all/blob/master/index.ts. Lets you match functions to make it impossible to accidentally promisify something that doesn't actually take a callback.

Kind of stalled because I was really just getting distracted while trying to solve promisifyAll, and I'm actually not sure that bit is properly doable :cry: